### PR TITLE
fix(host-service): heal trustd-degraded pty-daemons (gh x509 -26276)

### PR DESCRIPTION
**Pull request by Kitenite**
## Problem

`gh` (and other Go/Secure-Transport tools) fail inside Superset terminals with `tls: failed to verify certificate: x509: OSStatus -26276`, while working fine in a regular terminal.

**Root cause:** `-26276` = the process's macOS Mach bootstrap can't look up `com.apple.trustd` (the trust-evaluation daemon). The pty-daemon is detached and *adopted* across app/host-service restarts, so if it ever inherits a bootstrap that can't reach trustd (updater relaunch, a dead login-session port after logout/login), every terminal it spawns is broken — and adoption keeps the bad daemon alive indefinitely.

Reproduced deterministically with no signed build:
```
sandbox-exec -p '(deny mach-lookup (global-name "com.apple.trustd"))' gh api  →  x509: OSStatus -26276
```

Note: `SSL_CERT_FILE` is a **no-op** for `gh` here (Go uses the platform verifier), and Node/curl can't self-detect this (they don't use Secure Transport). Only `security verify-cert` exercises the same path.

## Fix (no `launchctl asuser`, no root)

host-service is a fresh `detached:false` child of the Aqua Electron app, so a plain daemon spawn from it always reaches trustd. So:

1. **Daemon self-probes** trustd at startup (`security verify-cert`) and reports `trustdHealthy` in the hello-ack.
2. **host-service**, on adopt: if the daemon reports trustd-degraded **and** host-service self-probes healthy → kill it and respawn fresh (which lands in a healthy bootstrap). If host-service is itself degraded, adopt and let the next healthy launch heal it.
3. **Migration:** daemon version `0.2.4 → 0.3.0` — existing daemons smooth-update onto the trustd-aware binary (sessions preserved via fd-handoff), then degraded ones self-report and get respawned on a later adopt. No forced mass session loss.

The probe is conservative: only a clean non-zero exit means degraded (errors/timeouts → healthy), so a flaky probe never triggers a session-destroying respawn.

## Test Plan

- [x] Reproduced `-26276` and validated `security verify-cert` as the discriminator (agrees with the real `gh` symptom)
- [x] End-to-end: real daemon spawned under a trustd-deny sandbox is detected (`pty_daemon_trustd_degraded_respawn`) and replaced with a healthy one
- [x] Probe unit tests (mutation-checked): 8
- [x] pty-daemon unit 65, integration 48
- [x] host-service unit 47, integration 18
- [x] typecheck + biome clean across pty-daemon, host-service, desktop

## Note

Supersedes the `detached:false`-on-darwin approach in #5429 — that rests on the (empirically refuted) premise that detachment severs the bootstrap, and it removes PTY signal isolation.

---

<a href="https://stagereview.app/superset-sh/superset/pull/5450">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Added a macOS reachability probe for the system trust service and exposed the result during the daemon startup/handshake.
  * Daemon servers can now report and update this “trustd healthy” flag after binding (including during handoff).

* **Bug Fixes**
  * If an adopted daemon reports it can’t reach the trust service while the host can, it’s automatically terminated and replaced with a fresh daemon.

* **Tests**
  * Added coverage for the trustd probe behavior and edge cases.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 4 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Implement macOS trustd reachability probe](https://stagereview.app/superset-sh/superset/pull/5450/chapters/1) |
| 2 | [Update pty-daemon protocol and version](https://stagereview.app/superset-sh/superset/pull/5450/chapters/2) |
| 3 | [Integrate trustd probe into pty-daemon](https://stagereview.app/superset-sh/superset/pull/5450/chapters/3) |
| 4 | [Heal degraded daemons in DaemonSupervisor](https://stagereview.app/superset-sh/superset/pull/5450/chapters/4) |

<a href="https://stagereview.app/superset-sh/superset/pull/5450">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 90edeb8830d32966225e847d31a64b7836af5696 on Jul 4, 2026 8:11am UTC.</sub>

File: packages/pty-daemon/src/main.ts [hunk 1]
```diff
@@ -136,7 +138,11 @@ async function runHandoffReceiver(): Promise<void> {
 		return;
 	}
 	log(`read snapshot: sessions=${snapshot.sessions.length}`);
-	const server = new Server({ socketPath, daemonVersion });
+	const server = new Server({
+		socketPath,
+		daemonVersion,
+		trustdHealthy: probeTrustdHealthy(),
+	});
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/pty-daemon/src/main.ts]: Fixed in 90edeb8: the probe now runs AFTER `server.listen()` (fresh) and AFTER the upgrade-ack + bind (handoff). `Server.trustdHealthy` is settable post-listen, so a slow/wedged `security` can no longer delay the ack or the socket coming up. Until the value lands, hello-ack omits it and the supervisor treats that as unknown (no respawn).

File: packages/pty-daemon/src/main.ts [hunk 2]
```diff
@@ -75,6 +76,7 @@ async function runFresh(): Promise<void> {
 	const server = new Server({
 		socketPath: args.socket,
 		daemonVersion,
+		trustdHealthy: probeTrustdHealthy(),
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/pty-daemon/src/main.ts]: Fixed in 90edeb8 — probe moved off the critical path (after bind + after upgrade-ack), so a wedged `security` no longer delays ack or socket-ready. That also removes the load-bearing timing dependency the comment flagged.

File: packages/pty-daemon/src/trustd-probe.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+			spawnSync(cmd, args, { timeout: PROBE_TIMEOUT_MS }));
+
+	let certPath: string | undefined;
+	try {
+		const bundle = deps.readBundle
+			? deps.readBundle()
+			: fs.readFileSync(SYSTEM_CERT_BUNDLE, "utf8");
+		const begin = bundle.indexOf(BEGIN);
+		const end = bundle.indexOf(END);
+		if (begin === -1 || end === -1 || end < begin) return true;
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/pty-daemon/src/trustd-probe.ts]: Applied in 90edeb8 (`bundle.indexOf(END, begin)`), plus a test with a leading `BEGIN TRUSTED CERTIFICATE` block.

File: packages/pty-daemon/src/trustd-probe.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		((cmd: string, args: string[]) =>
+			spawnSync(cmd, args, { timeout: PROBE_TIMEOUT_MS }));
+
+	let certPath: string | undefined;
+	try {
+		const bundle = deps.readBundle
+			? deps.readBundle()
+			: fs.readFileSync(SYSTEM_CERT_BUNDLE, "utf8");
+		const begin = bundle.indexOf(BEGIN);
+		const end = bundle.indexOf(END);
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/pty-daemon/src/trustd-probe.ts]: Fixed in 90edeb8 via `bundle.indexOf(END, begin)`; added a mixed-PEM-header test case.

File: packages/pty-daemon/src/trustd-probe.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		if (begin === -1 || end === -1 || end < begin) return true;
+		// A cert from the trust store verifies cleanly WHEN trustd is reachable,
+		// so a non-zero exit isolates "trustd unreachable" from "bad cert".
+		const cert = `${bundle.slice(begin, end + END.length)}\n`;
+		certPath = path.join(
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/pty-daemon/src/trustd-probe.ts]: Fixed in 90edeb8: the cert is now written into an `fs.mkdtempSync` 0700 dir (unpredictable path, fresh each run), so there's no predictable target to pre-seed as a symlink and no cross-run collision. Dir is rm'd in `finally`.

**Comment by Kitenite** [comment]
Superseded by #6302 (merged as c9cd91eec) — same fix revived and completed: async probe (daemon-loop ratchet), desktop entry shim coverage, heal gates + late-report heal, and regression tests.
